### PR TITLE
os/bluestore: set upper and lower bounds on rocksdb omap iterators

### PR DESCRIPTION
**Pull request by cfsnyder**
Limits RocksDB omap Seek operations to the relevant key range of the object's omap.
This prevents RocksDB from unnecessarily iterating over delete range tombstones in
irrelevant omap CF shards. Avoids extreme performance degradation commonly caused
by tombstones generated from RGW bucket resharding cleanup.  Also prefer CFIteratorImpl
over ShardMergeIteratorImpl when we can determine that all keys within specified
IteratorBounds must be in a single CF.

Fixes: https://tracker.ceph.com/issues/55324
Signed-off-by: Cory Snyder <csnyder@iland.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

### CODE REVIEW

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -2729,16 +2741,25 @@ class ShardMergeIteratorImpl : public KeyValueDB::IteratorImpl {
...
+				  const std::vector<rocksdb::ColumnFamilyHandle*>& shards,
+                  const KeyValueDB::IteratorBounds bounds)
+    : db(db), keyless(db->comparator), prefix(prefix), bounds(bounds)
   {
     iters.reserve(shards.size());
     for (auto& s : shards) {
-      iters.push_back(db->db->NewIterator(rocksdb::ReadOptions(), s));
+      auto options = rocksdb::ReadOptions();
+      if (bounds.upper_bound) {
+        options.iterate_upper_bound = new rocksdb::Slice(*this->bounds.upper_bound);
```
Comment (cbodley) [inline src/kv/RocksDBStore.cc:2757]: does rocksdb free this memory? i don't think it does, so we need to manage the memory ourselves

i pushed a commit https://github.com/cbodley/ceph/commit/9baa3db8daca52a95b0d817db7de489e3c53e4ad as an example (based on your earlier [wip-cfsnyder-rgw-updates-pacific ](https://github.com/ceph/ceph-ci/commits/wip-cfsnyder-rgw-updates-pacific) branch), that constructs these slices as member variables instead

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -12146,7 +12152,15 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
...
-  KeyValueDB::Iterator it = db->get_iterator(o->get_omap_prefix());
+  auto bounds = KeyValueDB::IteratorBounds();
+  if (o->onode.has_omap()) {
+    std::string lower_bound, upper_bound;
+    o->get_omap_key(string(), &lower_bound);
+    o->get_omap_tail(&upper_bound);
+    bounds.lower_bound = lower_bound;
+    bounds.upper_bound = upper_bound;
+  }
+  KeyValueDB::Iterator it = db->get_iterator(o->get_omap_prefix(), 0, bounds);
```
Comment (cbodley) [inline src/os/bluestore/BlueStore.cc:12163]: there are a several places that we're copying strings unnecessarily. these omap keys may tend to be large, so each copy would mean an extra memory allocation

prefer to move strings if you're done using them:
```c++
    bounds.lower_bound = std::move(lower_bound);
    bounds.upper_bound = std::move(upper_bound);
```
this goes for `KeyValueDB::IteratorBounds` too, which gets passed to functions by value:
```c++
  KeyValueDB::Iterator it = db->get_iterator(o->get_omap_prefix(), 0, std::move(bounds));
```

File: src/kv/RocksDBStore.cc (hunk 1 above)
Comment (cfsnyder) [reply src/kv/RocksDBStore.cc:2757]: I had originally stored pointers as member variables and was deleting them in the destructor, but was getting runtime errors from the allocator. I had assumed at the time that rocksdb was freeing this memory and the allocator was complaining due to a double free - but it wasn't clear. I'm testing your branch on my test cluster now and will definitely update to use your approach if things behave properly.

Comment (cbodley) [reply src/kv/RocksDBStore.cc:2757]: we chatted a bit about this on irc, and agreed that the `.value_or("")` part there is wrong because it returns a temporary std::string. i think we'd need a helper function like this:
```c++
static rocksdb::Slice make_slice(const std::optional<std::string>& bound) {
  if (bound) {
    return {*bound};
  } else {
    return {};
  }
}
```
so we can construct the member Slices like:
```c++
      bounds(std::move(bounds_)),
      iterate_lower_bound(make_slice(bounds.lower_bound)),
      iterate_upper_bound(make_slice(bounds.upper_bound))
```

**Comment by cfsnyder** [comment]
jenkins test api

**Comment by cfsnyder** [comment]
jenkins test api

File: src/kv/RocksDBStore.cc [hunk 3]
```diff
@@ -646,10 +655,36 @@ rocksdb::ColumnFamilyHandle *RocksDBStore::get_cf_handle(const std::string& pref
...
+      return get_key_cf(iter->second, key, keylen);
+    }
+  }
+}
+
+/**
+ * If the specified IteratorBounds arg has both an upper and a lower bound defined, and they have equal placement hash
+ * strings, we can be sure that the entire iteration range exists in a single CF. In that case, we return the relevant
+ * CF handle. In all other cases, we return a nullptr to indicate that the specified bounds cannot necessarily be mapped
+ * to a single CF.
```
Comment (baergj) [inline src/kv/RocksDBStore.cc:667]: This is indeed elegant! Are there specific things that users can do in their config to ensure this is the case most of the time?

Comment (baergj) [inline src/kv/RocksDBStore.cc:667]: This is indeed elegant! Are there specific things that users can do in their config to ensure this is the case most of the time?

Comment (cfsnyder) [reply src/kv/RocksDBStore.cc:667]: It is always the case for omap, as long as the `bluestore_rocksdb_cfs` is set to the default value (or is using an alternative hash constraint that still maps all keys for an individual omap object to the same CF).

File: src/kv/RocksDBStore.h [hunk 4]
```diff
@@ -64,6 +64,14 @@ namespace rocksdb{
 
 extern rocksdb::Logger *create_rocksdb_ceph_logger();
 
+static rocksdb::Slice make_slice(const std::optional<std::string>& bound) {
+    if (bound) {
```
Comment (cbodley) [inline src/kv/RocksDBStore.h:68]: at namespace-level, `static` means that each translation unit that includes "RocksDBStore.h" would get a separate copy of the function. prefer `inline` here so they share a single definition

(were this inside a class, the `static` would have a different meaning - static member functions would be shared instead of duplicated)

also nit: use two-space indentation in the function body

File: src/kv/RocksDBStore.h [hunk 5]
```diff
@@ -341,9 +353,29 @@ class RocksDBStore : public KeyValueDB {
...
+    const rocksdb::Slice iterate_lower_bound;
+    const rocksdb::Slice iterate_upper_bound;
   public:
-    explicit RocksDBWholeSpaceIteratorImpl(rocksdb::Iterator *iter) :
-      dbiter(iter) { }
+    explicit RocksDBWholeSpaceIteratorImpl(const RocksDBStore* db,
+                                           rocksdb::ColumnFamilyHandle* cf,
+                                           const KeyValueDB::IteratorOpts opts,
+                                           const KeyValueDB::IteratorBounds bounds_) :
+      bounds(std::move(bounds_)),
```
Comment (cbodley) [inline src/kv/RocksDBStore.h:364]: note that move semantics require non-const input - we want `bounds` to 'steal' the memory owned by `bounds_`, and it can't do that safely without modifying `bounds_` - otherwise its destructor would free the memory that `bounds` is still referencing

another way to look at it, is that we're trying to call the compiler-generated move constructor of `IteratorBounds`:
```c++
  IteratorBounds(IteratorBounds&& other) = default; // move constructor
  IteratorBounds(const IteratorBounds& other) = default; // copy constructor
```
`std::move()` is just a cast to rvalue-reference `&&`, so given `const IteratorBounds bounds_`, the return type of `std::move(bounds_)` would be `const IteratorBounds&&`. this doesn't match the move constructor overload, so would call the copy constructor instead

tldr: remove the `const` from `bounds_` here, and do the same for `ShardMergeIteratorImpl` and `CFIteratorImpl`

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -12146,7 +12152,15 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
...
   o->flush();
   dout(10) << __func__ << " has_omap = " << (int)o->onode.has_omap() <<dendl;
-  KeyValueDB::Iterator it = db->get_iterator(o->get_omap_prefix());
+  auto bounds = KeyValueDB::IteratorBounds();
+  if (o->onode.has_omap()) {
+    std::string lower_bound, upper_bound;
+    o->get_omap_key(string(), &lower_bound);
+    o->get_omap_tail(&upper_bound);
+    bounds.lower_bound = lower_bound;
+    bounds.upper_bound = upper_bound;
```
Comment (cbodley) [inline src/os/bluestore/BlueStore.cc:12161]: can use move-assignment here, because lower_bound and upper_bound are about to go out of scope

**Review by cbodley — COMMENTED** [review]
👍

File: src/kv/RocksDBStore.cc (hunk 3 above)
Comment (baergj) [reply src/kv/RocksDBStore.cc:667]: OK, great - just wondering if it's worth throwing notes in the config docs so that users know what the optimized path is.

File: src/kv/RocksDBStore.h (hunk 4 above)
Comment (cfsnyder) [reply src/kv/RocksDBStore.h]: thanks for the explanation, @cbodley !

File: src/kv/RocksDBStore.h (hunk 5 above)
Comment (cfsnyder) [reply src/kv/RocksDBStore.h]: excellent explanation, thanks again @cbodley !

File: src/kv/RocksDBStore.cc [hunk 7]
```diff
@@ -2933,23 +3007,19 @@ rocksdb::Iterator* RocksDBStore::new_shard_iterator(rocksdb::ColumnFamilyHandle*
...
-RocksDBStore::WholeSpaceIterator RocksDBStore::get_wholespace_iterator(IteratorOpts opts)
+RocksDBStore::WholeSpaceIterator RocksDBStore::get_wholespace_iterator(IteratorOpts opts, IteratorBounds bounds)
 {
   if (cf_handles.size() == 0) {
-    rocksdb::ReadOptions opt = rocksdb::ReadOptions();
-    if (opts & ITERATOR_NOCACHE)
-      opt.fill_cache=false;
     return std::make_shared<RocksDBWholeSpaceIteratorImpl>(
-      db->NewIterator(opt, default_cf));
+      this, default_cf, opts, bounds);
```
Comment (cbodley) [inline src/kv/RocksDBStore.cc:3014]: can `std::move(bounds)` here too?

**Review by cbodley — COMMENTED** [review]
otherwise looks great. i'll defer to bluestore folks for final approval

**Comment by cfsnyder** [comment]
@cbodley should I tag someone specific from the bluestore side for review?

**Comment by ifed01** [comment]
Generally this looks good to me. I have a concern though: may be it's better to make this feature's enablement configurable given that we're planning to backport it ASAP and hence can provide quite a limited testing.
@cfsnyder @neha-ojha @aclamk @markhpc - what do you think?

**Review by markhpc — APPROVED** [review]
superficially looks good to me, I trust @ifed01 and @cbodley's take on the code.  Once the option to disable the behavior gets added (default on) let's get this tested and merged.

**Comment by ljflores** [comment]
@yuriw ^

File: src/kv/RocksDBStore.h [hunk 8]
```diff
@@ -119,8 +128,11 @@ class RocksDBStore : public KeyValueDB {
   void add_column_family(const std::string& cf_name, uint32_t hash_l, uint32_t hash_h,
 			 size_t shard_idx, rocksdb::ColumnFamilyHandle *handle);
   bool is_column_family(const std::string& prefix);
+  std::string_view get_key_hash_view(const prefix_shards& shards, const char* key, const size_t keylen);
+  rocksdb::ColumnFamilyHandle *get_key_cf(const prefix_shards& shards, const char* key, const size_t keylen);
   rocksdb::ColumnFamilyHandle *get_cf_handle(const std::string& prefix, const std::string& key);
   rocksdb::ColumnFamilyHandle *get_cf_handle(const std::string& prefix, const char* key, size_t keylen);
+  rocksdb::ColumnFamilyHandle *get_cf_handle(const std::string& prefix, const IteratorBounds& bounds);
```
Comment (aclamk) [inline src/kv/RocksDBStore.h:135]: Love that this new function is made similar to existing ones.

File: src/kv/RocksDBStore.cc [hunk 9]
```diff
@@ -2194,10 +2229,29 @@ class CFIteratorImpl : public KeyValueDB::IteratorImpl {
...
+  explicit CFIteratorImpl(const RocksDBStore* db,
+                          const std::string& p,
+                          rocksdb::ColumnFamilyHandle* cf,
+                          KeyValueDB::IteratorBounds bounds_)
+    : prefix(p), bounds(std::move(bounds_)),
+      iterate_lower_bound(make_slice(bounds.lower_bound)),
+      iterate_upper_bound(make_slice(bounds.upper_bound))
+      {
+      auto options = rocksdb::ReadOptions();
+      if (db->cct->_conf.get_val<bool>("osd_rocksdb_iterator_bounds_enabled")) {
```
Comment (cbodley) [inline src/kv/RocksDBStore.cc:2245]: `get_val()` unfortunately requires a global mutex that we probably don't want to pay for here. if you add this line to the config option:
```
with_legacy: true
```
you can access this configurable directly without locking:
```c++
if (db->cct->_conf->osd_rocksdb_iterator_bounds_enabled) {
```

Comment (cfsnyder) [reply src/kv/RocksDBStore.cc:2245]: thanks @cbodley, I was wondering what affect "with_legacy" had. Updating it now!

**Review by cbodley — APPROVED** [review]

**Comment by ljflores** [comment]
@yuriw extra commit has been added ^

File: src/kv/RocksDBStore.cc [hunk 10]
```diff
@@ -646,10 +655,36 @@ rocksdb::ColumnFamilyHandle *RocksDBStore::get_cf_handle(const std::string& pref
...
+ * If the specified IteratorBounds arg has both an upper and a lower bound defined, and they have equal placement hash
+ * strings, we can be sure that the entire iteration range exists in a single CF. In that case, we return the relevant
+ * CF handle. In all other cases, we return a nullptr to indicate that the specified bounds cannot necessarily be mapped
+ * to a single CF.
+ */
+rocksdb::ColumnFamilyHandle *RocksDBStore::get_cf_handle(const std::string& prefix, const IteratorBounds& bounds) {
+  if (!bounds.lower_bound || !bounds.upper_bound) {
+    return nullptr;
+  }
+  auto iter = cf_handles.find(prefix);
```
Comment (aclamk) [inline src/kv/RocksDBStore.cc:673]: We invoke this only from `RocksDBStore::get_iterator`.
We should not implement here cases that should not happen here.
So instead of elaborate logic we should do:
```
ceph_assert(iter != cf_handles.end());
ceph_assert(iter->second.handles.size() != 1);
```

File: src/kv/RocksDBStore.cc (hunk 3 above)
Comment (aclamk) [inline src/kv/RocksDBStore.cc:667]: The comment here is a bit wrong.
It is of course true that for restricting search for one column family we must make sure that no key can squeeze between lower and upper bound and have different hash. 

To be sure that all keys in range [lower_bound .. upper_bound] belong to one shard, following must be true:
```
0000000000000000111111111
0123456789abcdef012345678
lower_bound_value
upper_bound_value
  ^--hash--^
```
1) lower_bound.size() >= hash_h
2) upper_bound.size() >= hash_h
3) lower_bound[0..hash_h] == upper_bound[0..hash_h]

Important part here is, that first hash_l characters cannot be ignored, as it would trigger as going to one shard:
```
0000000000000000111111111
0123456789abcdef012345678
ABC_bound_value_lower
DEF_bound_value_upper
    ^--hash--^
```

File: src/kv/RocksDBStore.cc [hunk 11]
```diff
@@ -646,10 +655,36 @@ rocksdb::ColumnFamilyHandle *RocksDBStore::get_cf_handle(const std::string& pref
...
+  if (iter == cf_handles.end() || iter->second.hash_l != 0) {
+    return nullptr;
```
Comment (aclamk) [inline src/kv/RocksDBStore.cc:675]: I see here a check for hash_l==0.
This makes the implementation valid.

File: src/kv/RocksDBStore.cc (hunk 10 above)
Comment (cfsnyder) [reply src/kv/RocksDBStore.cc:673]: I've added another commit that updates this code with your suggestion!

File: src/kv/RocksDBStore.cc (hunk 3 above)
Comment (cfsnyder) [reply src/kv/RocksDBStore.cc:667]: I've updated the comment to include a note about that bit

**Comment by aclamk** [comment]
jenkins make check

**Comment by aclamk** [comment]
jenkins test make check

**Review by aclamk — APPROVED** [review]
Excellent improvement!

**Comment by ifed01** [comment]
jenkins test make check

**Review by ifed01 — APPROVED** [review]

**Comment by ljflores** [comment]
@cfsnyder it's going through some final testing. We plan to get this merged today once the testing is validated.

**Comment by ljflores** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2022-04-22_13:56:48-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/yuriw-2022-04-23_16:21:59-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/lflores-2022-04-25_16:23:25-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/

Failures, unrelated:
    https://tracker.ceph.com/issues/55419
    https://tracker.ceph.com/issues/55429
    https://tracker.ceph.com/issues/54458

Details:
    1. cephtool/test.sh: failure on blocklist testing - Ceph - RADOS
    2. mgr/dashboard: AttributeError: 'NoneType' object has no attribute 'group' - Ceph - Mgr - Dashboard
    3. osd-scrub-snaps.sh: TEST_scrub_snaps failed due to malformed log message - Ceph - RADOS

**Comment by neha-ojha** [comment]
jenkins test make check

**Review by neha-ojha — APPROVED** [review]
Thanks @cfsnyder!

**Comment by neha-ojha** [comment]
jenkins test make check

**Comment by neha-ojha** [comment]
jenkins test make check

**Comment by idryomov** [comment]
This change broke upgrades, from octopus to pacific at least.  Reverting https://github.com/ceph/ceph/pull/45963 backport from pacific restores omap functionality.  See https://tracker.ceph.com/issues/55444 for details.
